File: 684-redundant-connection/redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        # create adjacency list
        adj = defaultdict(lambda: [])
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])
        visited = set()

         # Do DFS
        def dfs(node: int, history: List[int], parent:int) -> (list[int], int):
            print(history)
            if node in visited:
                return (history,  node)
            visited.add(node)
            starting_node = -1
            history.append(node)
            for child in adj[node]:
                if child == parent: continue
                (sol, starting_node) = dfs(child, history, node)
                if starting_node != -1:
                    return (sol, starting_node)
            history.pop()
            return ([], -1)

        # trimmed history to where the loop node starts
        s = 1
        history, loop_start = dfs(s, [], -1)
        start_pos = 0
        for i in range(len(history)):
            if history[i] == loop_start:
                start_pos = i
                break
        history = history[start_pos:]
        
        loop_edges = set()
        for i in range(len(history)-1):
            loop_edges.add((history[i], history[i+1]))
            loop_edges.add((history[i+1], history[i]))
        loop_edges.add((history[0], history[-1]))
        loop_edges.add((history[-1], history[0]))
            
        for i in range(len(edges)-1, -1, -1):
            cur_edge = edges[i]
            if (cur_edge[0], cur_edge[1]) in loop_edges:
                return (cur_edge[0], cur_edge[1])
        return -1

File: 684-redundant-connection/redundant-connection.cpp
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int max_node = 0;
        unordered_set<int> visited;
        for(const auto& edge:edges){
            max_node = max(max_node, edge[0]);
            max_node = max(max_node, edge[1]);
        }

        vector<vector<int>> graph(max_node + 1, vector<int>());
        for(const auto& edge:edges){
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        vector<int> parent(max_node+1, -1);
        int u = -1;
        int v = -1;
        dfs(graph, visited, parent, 1, -1, &u, &v);

        const auto pair_hash = [](const pair<int, int> p) {
            return hash<int>{}(p.first) ^ hash<int>{}(p.second);
        };
        unordered_set<pair<int, int>, decltype(pair_hash)> cycle_nodes;
        int head = u;
        while(head != v){
            cycle_nodes.insert({head, parent[head]});
            cycle_nodes.insert({parent[head], head});
            head = parent[head];
        }
        cycle_nodes.insert({u, v});
        cycle_nodes.insert({v, u});

        reverse(edges.begin(), edges.end());
        
        for(const auto& edge: edges){
            if(cycle_nodes.contains({edge[0], edge[1]})){
                return vector<int> {edge[0], edge[1]};
            }
        }
        return vector<int>();
    }

    void dfs(vector<vector<int>>& graph,unordered_set<int>& visited,vector<int>& p, int node, int parent, int * u, int * v) {
        if(*u != -1) return; 
        if(visited.contains(node)){
            *v = node;
            return;
        }
        visited.insert(node);
        p[node] = parent;
        for(int child_node: graph[node]){
            if(child_node == parent) continue;
            dfs(graph, visited,p, child_node, node, u, v);
            if(*v != -1 && *u == -1){
                *u = node;
                return;
            }
        }
    }
};